iroh-gossip/src/proto/hyparview.rs

```diff
@@ -0,0 +1,707 @@
+        // "A node that receives a join request will start by adding the new
+        // node to its active view, even if it has to drop a random node from it. (6)"
+        self.add_active(peer, data.clone(), Priority::High, now, io);
+        // "The contact node c will then send to all other nodes in its active view a ForwardJoin
+        // request containing the new node identifier. Associated to the join procedure,
+        // there are two configuration parameters, named Active Random Walk Length (ARWL),
+        // that specifies the maximum number of hops a ForwardJoin request is propagated,
+        // and Passive Random Walk Length (PRWL), that specifies at which point in the walk the node
+        // is inserted in a passive view. To use these parameters, the ForwardJoin request carries
+        // a “time to live” field that is initially set to ARWL and decreased at every hop. (7)"
+        let ttl = self.config.active_random_walk_length;
+        let peer_info = PeerInfo { id: peer, data };
+        for node in self.active_view.iter_without(&peer) {
+            let message = Message::ForwardJoin(ForwardJoin {
+                peer: peer_info.clone(),
+                ttl,
+            });
+            io.push(OutEvent::SendMessage(*node, message));
+        }
+    }
+
+    fn on_forward_join(
+        &mut self,
+        sender: PA,
+        message: ForwardJoin<PA>,
+        now: Instant,
+        io: &mut impl IO<PA>,
+    ) {
+        // "i) If the time to live is equal to zero or if the number of nodes in p’s active view is equal to one,
+        // it will add the new node to its active view (7)"
+        if message.ttl.expired() || self.active_view.len() <= 1 {
+            self.add_active(
+                message.peer.id,
+                message.peer.data.clone(),
+                Priority::High,
+                now,
+                io,
+            );
+        }
+        // "ii) If the time to live is equal to PRWL, p will insert the new node into its passive view"
+        else if message.ttl == self.config.passive_random_walk_length {
+            self.add_passive(message.peer.id, message.peer.data.clone(), io);
+        }
+        // "iii) The time to live field is decremented."
+        // "iv) If, at this point, n has not been inserted
+        // in p’s active view, p will forward the request to a random node in its active view
+        // (different from the one from which the request was received)."
+        if !self.active_view.contains(&message.peer.id) {
+            match self
+                .active_view
+                .pick_random_without(&[&sender], &mut self.rng)
+            {
+                None => {
+                    // TODO: I think this is unreachable!() but will have to check, maybe decrease
+                    // to warn
+                    unreachable!("no peers in active view but also did not add this node on forward join, this should not happen");
```

Comment:
this looks unreachable to me too, but I would suggest instead:
```suggestion
                    unreachable!("if the peer was not added, there are at least two peers in our active view.");
```